File: app/src/main/java/ru/org/adons/cblock/app/ApplicationModule.java
package ru.org.adons.cblock.app;

import android.content.Context;

import com.squareup.sqlbrite.BriteContentResolver;
import com.squareup.sqlbrite.SqlBrite;

import org.greenrobot.greendao.database.Database;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import ru.org.adons.cblock.model.DaoMaster;
import ru.org.adons.cblock.model.DaoSession;
import rx.schedulers.Schedulers;

/**
 * Provide application scope dependencies (Context, DaoSession, Preferences)
 */
@Module
class ApplicationModule {

    private final Context context;

    ApplicationModule(Context context) {
        this.context = context;
    }

    @Singleton
    @Provides
    Context provideContext() {
        return context;
    }

    @Singleton
    @Provides
    Preferences providePreferences() {
        return new Preferences(context);
    }

    @Singleton
    @Provides
    BriteContentResolver provideContentResolver(SqlBrite sqlBrite) {
        return sqlBrite.wrapContentProvider(context.getContentResolver(), Schedulers.immediate());
    }

    @Singleton
    @Provides
    SqlBrite provideSqlBrite() {
        return new SqlBrite.Builder().build();
    }

    @Singleton
    @Provides
    DaoSession provideDaoSession() {
        DaoMaster.DevOpenHelper helper = new DaoMaster.DevOpenHelper(context, "cblock-db");
        Database db = helper.getWritableDb();
        return new DaoMaster(db).newSession();
    }

}

File: app/src/main/java/ru/org/adons/cblock/app/ApplicationModule.kt
package ru.org.adons.cblock.app

import android.content.Context
import com.squareup.sqlbrite.BriteContentResolver
import com.squareup.sqlbrite.SqlBrite
import dagger.Module
import dagger.Provides
import ru.org.adons.cblock.data.DaoMaster
import ru.org.adons.cblock.data.DaoSession
import rx.schedulers.Schedulers
import javax.inject.Singleton

/**
 * Provide application scope dependencies (Context, DaoSession, Preferences)
 */
@Module
internal class ApplicationModule(private val context: Context) {

    @Singleton
    @Provides
    fun provideContext(): Context {
        return context
    }

    @Singleton
    @Provides
    fun provideContentResolver(sqlBrite: SqlBrite): BriteContentResolver {
        return sqlBrite.wrapContentProvider(context.contentResolver, Schedulers.immediate())
    }

    @Singleton
    @Provides
    fun provideSqlBrite(): SqlBrite {
        return SqlBrite.Builder().build()
    }

    @Singleton
    @Provides
    fun provideDaoSession(): DaoSession {
        val helper = DaoMaster.DevOpenHelper(context, "cblock-db")
        val db = helper.writableDb
        return DaoMaster(db).newSession()
    }

}
